beacon-chain/sync/rpc_data_column_sidecars_by_range.go

```diff
@@ -0,0 +1,218 @@
+		if maxRequestDataColumnSidecars == 0 {
+			log.WithField("initialQuota", beaconConfig.MaxRequestDataColumnSidecars).Debug("Reached quota for data column sidecars by range request")
+			break
+		}
+	}
+
+	if err := batch.error(); err != nil {
+		log.WithError(err).Debug("error in DataColumnSidecarsByRange batch")
+
+		// If we hit a rate limit, the error response has already been written, and the stream is already closed.
+		if !errors.Is(err, p2ptypes.ErrRateLimited) {
+			s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		}
+
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	closeStream(stream, log)
+	return nil
+}
+
+func (s *Service) streamDataColumnBatch(ctx context.Context, batch blockBatch, quota uint64, wantedDataColumnIndices []uint64, stream libp2pcore.Stream) (uint64, error) {
+	_, span := trace.StartSpan(ctx, "sync.streamDataColumnBatch")
+	defer span.End()
+
+	// Defensive check to guard against underflow.
+	if quota == 0 {
+		return 0, nil
+	}
+
+	// Loop over the blocks in the batch.
+	for _, block := range batch.canonical() {
+		// Get the block blockRoot.
+		blockRoot := block.Root()
+
+		// Retrieve the data column sidecars from the store.
+		verifiedRODataColumns, err := s.cfg.dataColumnStorage.Get(blockRoot, wantedDataColumnIndices)
+		if err != nil {
+			s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+			return quota, errors.Wrapf(err, "get data column sidecars: block root %#x", blockRoot)
+		}
+
+		// Write the retrieved sidecars to the stream.
+		for _, verifiedRODataColumn := range verifiedRODataColumns {
+			sidecar := verifiedRODataColumn.DataColumnSidecar
+			SetStreamWriteDeadline(stream, defaultWriteDuration)
+
+			if err := WriteDataColumnSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sidecar); err != nil {
+				s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+				tracing.AnnotateError(span, err)
+				return quota, errors.Wrap(err, "write data column sidecar chunk")
+			}
+
+			s.rateLimiter.add(stream, rateLimitingAmount)
+			quota -= 1
+
+			// Stop streaming results once the quota of writes for the request is consumed.
+			if quota == 0 {
+				return 0, nil
+			}
+		}
+	}
+
+	return quota, nil
+}
+
+func validateDataColumnsByRange(request *pb.DataColumnSidecarsByRangeRequest, currentSlot primitives.Slot) (*rangeParams, error) {
+	startSlot, count := request.StartSlot, request.Count
+
+	if count == 0 {
+		return nil, errors.Wrap(p2ptypes.ErrInvalidRequest, "invalid request count parameter")
+	}
+
+	endSlot, err := request.StartSlot.SafeAdd(count - 1)
+	if err != nil {
+		return nil, errors.Wrap(p2ptypes.ErrInvalidRequest, "overflow start + count -1")
+	}
+
+	// Peers may overshoot the current slot when in initial sync,
+	// so we don't want to penalize them by treating the request as an error.
+	if startSlot > currentSlot {
+		return nil, nil
```

Comment:
wondering if it's good practice or even helpful for us to add logs when it returns nil,nil